### PR TITLE
chore: migrate SDK to a generic OrderBy [MLG-1056]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

introduces a new `OrderBy` and deprecates existing class-specific classes (`CheckpointOrderBy`, `TrialOrderBy`, etc.)

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/common/experimental/_util.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+from determined.common.api import bindings
+
+
+class OrderBy(enum.Enum):
+    """
+    Specifies whether a sorted list of objects should be in ascending or
+    descending order.
+    """
+
+    ASCENDING = bindings.v1OrderBy.ASC.value
```
Comment (wes-turner) [inline harness/determined/common/experimental/_util.py:12]: I don't know if this is standard or "best practices", but I think this is a nice touch.

File: harness/determined/common/experimental/_util.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import enum
+
+from determined.common.api import bindings
+
+
+class OrderBy(enum.Enum):
+    """
```
Comment (wes-turner) [inline harness/determined/common/experimental/_util.py:7]: nit: small request to follow [style guide format](https://google.github.io/styleguide/pyguide.html#384-classes) for this docstring.

File: harness/determined/common/experimental/_util.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    Specifies whether a sorted list of objects should be in ascending or
+    descending order.
+    """
+
+    ASCENDING = bindings.v1OrderBy.ASC.value
+    ASC = bindings.v1OrderBy.ASC.value
+    DESCENDING = bindings.v1OrderBy.DESC.value
+    DESC = bindings.v1OrderBy.DESC.value
+
+    def _to_bindings(self) -> bindings.v1OrderBy:
```
Comment (wes-turner) [inline harness/determined/common/experimental/_util.py:17]: Also a `_from_bindings`? Is that dumb?

File: harness/determined/common/experimental/experiment.py [hunk 4]
```diff
@@ -59,8 +60,18 @@ def _to_bindings(self) -> bindings.v1GetExperimentsRequestSortBy:
...
     """Specifies whether a sorted list of experiments should be in ascending or
     descending order.
+
+    This class is deprecated in favor of ``OrderBy`` and will be removed in a future
+    release.
     """
 
+    warnings.warn(
+        "'ExperimentOrderBy' is deprecated and will be removed in a future "
+        "release. Please use 'OrderBy' instead.",
```
Comment (wes-turner) [inline harness/determined/common/experimental/experiment.py:70]: Could you please add a module name to the `OrderBy`? Just something to help with the deprecation.

File: harness/determined/common/experimental/experiment.py [hunk 5]
```diff
@@ -59,8 +60,18 @@ def _to_bindings(self) -> bindings.v1GetExperimentsRequestSortBy:
 class ExperimentOrderBy(enum.Enum):
     """Specifies whether a sorted list of experiments should be in ascending or
     descending order.
+
+    This class is deprecated in favor of ``OrderBy`` and will be removed in a future
+    release.
     """
 
+    warnings.warn(
```
Comment (wes-turner) [inline harness/determined/common/experimental/experiment.py:68]: When does this actually get executed? I seem to be able to 

```python
experiment.ExperimentOrderBy.ASCENDING
```

without any warning (assume the same is true with `CheckpointOrderBy`).

Neither of my first couple ideas worked for printing a warning, so I'll just publish this comment for expedience and we'll come back as needed.

File: harness/determined/common/experimental/experiment.py [hunk 6]
```diff
@@ -9,6 +9,7 @@
 from determined.common import api
 from determined.common.api import bindings
 from determined.common.experimental import checkpoint, trial, workspace
+from determined.common.experimental._util import OrderBy
```
Comment (wes-turner) [inline harness/determined/common/experimental/experiment.py:12]: This violates [style guide](https://google.github.io/styleguide/pyguide.html#22-imports):
> Use import statements for packages and modules only, not for individual classes or functions.

We haven't been enforcing this broadly, but such enforcement is coming (I think there's an existing ticket).

File: harness/determined/common/experimental/__init__.py [hunk 7]
```diff
@@ -13,6 +13,8 @@
     ValidationMetrics,
 )
 from determined.common.experimental.project import Project
+from determined.common.experimental._util import OrderBy
```
Comment (wes-turner) [inline harness/determined/common/experimental/__init__.py:16]: bleh. This file shouldn't exist (who is supposed to use it?). I guess I prefer not adding to it? But I don't know, given that it does maybe it's good to keep updated.

File: harness/determined/common/experimental/_util.py (hunk 3 above)
Comment (azhou-determined) [reply harness/determined/common/experimental/_util.py:17]: thought about this at some point, but couldn't come up with a use case where you'd need to get `OrderBy` back from the REST API. don't want to preemptively add something that will just be dead code. we can add it if it's needed at some point. 🤷

File: harness/determined/common/experimental/experiment.py (hunk 6 above)
Comment (azhou-determined) [reply harness/determined/common/experimental/experiment.py:12]: discussed offline. imported as such to obfuscate _util. TODO and JIRA added.

File: harness/determined/common/experimental/_util.py (hunk 3 above)
Comment (wes-turner) [reply harness/determined/common/experimental/_util.py:17]: I admire that conservatism. I think you've made the right choice.

**Review by wes-turner — APPROVED** [review]
LGTM

**Pull request merged**